terraform/modules/aws-network-load-balancer/data.tf

```diff
@@ -0,0 +1,40 @@
+data "aws_vpc" "this" {
+  filter {
+    name   = "tag:Environment"
+    values = [ var.environment ]
+  }
+}
+
+data "aws_subnet_ids" "public_subnets" {
+  vpc_id = data.aws_vpc.this.id
+
+  filter {
+    name   = "tag:Environment"
+    values = [ var.environment ]
+  }
+
+  filter {
+    name   = "tag:Routability"
+    values = ["public"]
+  }
+}
+
+data "aws_instances" "nodes" {
+  filter {
+    name   = "tag:Environment"
+    values = [ var.environment ]
+  }
+
+  filter {
+    name   = "tag:Role"
+    values = [ var.target_role ]
+  }
+
+  instance_state_names = ["running"]
+}
+
+data "aws_instance" "nodes" {
+  for_each = { for k, v in data.aws_instances.nodes.ids : v => v }
+
+  instance_id = each.value
+}
```

Comment:
This discovers aws instances at the time the terraform script is run - meaning we need to run it again every time we add a node to the cluster.

Instead of doing this, can we just provision these instances as part of an autoscaling group, with a `aws_autoscaling_attachment` resource?

Reply:
You are right. Technically, that is absolutely possible. Though, it would limit the applicable scenarios - of course, with the advantage of removing a manual step.

The use case for this module - right now - are k8s master nodes. Putting those in an ACG would be debatable, I think. Anyway, extending the module to support should be a fairly simple task, which we can address as soon as we need such an approach.

To clarify on this, I'll add a note to the readme.

Reply:
IMO, this should not be a data statement. do we need to pass the list of instances into the module instead? It should be possible to do all of the terraform instantiation in one go.